File: backend/app/schemas/research_skill.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


ResearchSkillRubricStatus = Literal["draft", "active", "archived"]
ResearchSkillAssessmentStatus = Literal["draft", "submitted", "reviewed", "locked"]
ResearchSkillCriterionScope = Literal["overall", "phase"]
ResearchSkillScoreType = Literal["integer_scale", "number", "single_choice", "boolean", "text"]
ResearchSkillTargetType = Literal["overall", "phase_segment"]
ResearchSkillMutationAction = Literal[
    "created",
    "updated",
    "deleted",
    "unchanged",
    "evidence_created",
    "evidence_updated",
    "evidence_deleted",
]
ResearchSkillStatusMutationAction = Literal["submitted", "reopened"]
ResearchSkillValidationSeverity = Literal["error", "warning", "info"]
ResearchSkillValidationIssueType = Literal[
    "missing_required_score",
    "invalid_value",
    "na_not_allowed",
    "phase_score_without_segment",
    "segment_not_in_selected_phase_set",
    "criterion_not_applicable",
    "assessment_phase_set_missing",
    "assessment_phase_set_draft",
    "rubric_not_active",
    "inactive_criterion",
    "evidence_out_of_bounds",
    "incomplete_phase_criteria",
]


class ResearchSkillPhaseLabelResponse(BaseModel):
    id: int
    key: str
    name: str
    color: str

    model_config = {"from_attributes": True}


class ResearchSkillCriterionResponse(BaseModel):
    id: int
    rubric_id: int
    key: str
    name: str
    description: str | None = None
    scope: ResearchSkillCriterionScope
    score_type: ResearchSkillScoreType
    min_value: float | None = None
    max_value: float | None = None
    step: float | None = None
    options_json: list[dict[str, Any]] | None = None
    required: bool
    allow_na: bool
    weight: float | None = None
    display_order: int
    is_active: bool
    phase_label_ids: list[int] = Field(default_factory=list)
    phase_labels: list[ResearchSkillPhaseLabelResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchSkillRubricSummary(BaseModel):
    id: int
    name: str
    version: int
    description: str | None = None
    status: ResearchSkillRubricStatus
    phase_protocol_id: int | None = None
    created_by_id: int | None = None
    criterion_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchSkillRubricDetail(ResearchSkillRubricSummary):
    criteria: list[ResearchSkillCriterionResponse] = Field(default_factory=list)


class CreateResearchSkillRubricRequest(BaseModel):
    name: str = Field(max_length=255)
    version: int = Field(default=1, ge=1)
    description: str | None = None
    phase_protocol_id: int | None = None
    username: str | None = Field(default=None, max_length=255)


class UpdateResearchSkillRubricRequest(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    phase_protocol_id: int | None = None
    clear_phase_protocol: bool = False


class CloneResearchSkillRubricRequest(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None


class CreateResearchSkillCriterionRequest(BaseModel):
    key: str = Field(max_length=120)
    name: str = Field(max_length=255)
    description: str | None = None
    scope: ResearchSkillCriterionScope
    score_type: ResearchSkillScoreType
    min_value: float | None = None
    max_value: float | None = None
    step: float | None = None
    options_json: list[dict[str, Any]] | None = None
    required: bool = False
    allow_na: bool = False
    weight: float | None = None
    display_order: int = Field(ge=0)
    is_active: bool = True
    phase_label_ids: list[int] = Field(default_factory=list)


class UpdateResearchSkillCriterionRequest(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    min_value: float | None = None
    max_value: float | None = None
    step: float | None = None
    options_json: list[dict[str, Any]] | None = None
    required: bool | None = None
    allow_na: bool | None = None
    weight: float | None = None
    clear_weight: bool = False
    display_order: int | None = Field(default=None, ge=0)
    is_active: bool | None = None
    phase_label_ids: list[int] | None = None


class ActivateResearchSkillRubricResponse(BaseModel):
    rubric: ResearchSkillRubricDetail


class ArchiveResearchSkillRubricResponse(BaseModel):
    rubric: ResearchSkillRubricDetail


class ResearchSkillVideoSummary(BaseModel):
    id: int
    name: str
    fps: float | None = None
    frame_count: int
    duration_ms: int | None = None


class ResearchSkillPhaseSegmentSummary(BaseModel):
    id: int
    phase_label_id: int
    phase_key: str
    phase_name: str
    start_frame: int
    end_frame_exclusive: int | None = None


class ResearchSkillPhaseAnnotationSetSummary(BaseModel):
    id: int
    protocol_id: int
    status: str
    revision: int
    segments: list[ResearchSkillPhaseSegmentSummary] = Field(default_factory=list)


class ResearchSkillEvidenceResponse(BaseModel):
    id: int
    skill_score_id: int
    start_frame: int
    end_frame_exclusive: int | None = None
    comment: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchSkillScoreResponse(BaseModel):
    id: int
    assessment_id: int
    criterion_id: int
    criterion_key: str
    criterion_name: str
    scope: ResearchSkillCriterionScope
    score_type: ResearchSkillScoreType
    target_key: str
    phase_segment_id: int | None = None
    value: Any | None = None
    is_na: bool
    comment: str | None = None
    evidence: list[ResearchSkillEvidenceResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class ResearchSkillCompletionResponse(BaseModel):
    required_total: int
    required_completed: int
    overall_required_total: int
    overall_required_completed: int
    phase_required_total: int
    phase_required_completed: int
    completion_percent: float


class ResearchSkillAssessmentSummary(BaseModel):
    id: int
    video_id: int
    rubric_id: int
    rater_id: int
    phase_annotation_set_id: int | None = None
    status: ResearchSkillAssessmentStatus
    revision: int
    overall_comment: str | None = None
    submitted_at: datetime | None = None
    reviewed_at: datetime | None = None
    locked_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    rubric_name: str
    rubric_version: int
    rater_username: str
    score_count: int = 0

    model_config = {"from_attributes": True}


class ResearchSkillAssessmentDetail(ResearchSkillAssessmentSummary):
    video: ResearchSkillVideoSummary
    rubric: ResearchSkillRubricDetail
    phase_annotation_set: ResearchSkillPhaseAnnotationSetSummary | None = None
    scores: list[ResearchSkillScoreResponse] = Field(default_factory=list)
    completion: ResearchSkillCompletionResponse


class CreateResearchSkillAssessmentRequest(BaseModel):
    rubric_id: int
    username: str = Field(max_length=255)
    phase_annotation_set_id: int | None = None


class CreateResearchSkillAssessmentResponse(BaseModel):
    created: bool
    assessment: ResearchSkillAssessmentDetail


class UpdateResearchSkillAssessmentRequest(BaseModel):
    overall_comment: str | None = Field(default=None, max_length=10000)
    clear_overall_comment: bool = False
    phase_annotation_set_id: int | None = None
    clear_phase_annotation_set: bool = False
    expected_revision: int = Field(ge=1)


class UpsertResearchSkillScoreRequest(BaseModel):
    target_type: ResearchSkillTargetType
    phase_segment_id: int | None = None
    value: Any | None = None
    is_na: bool = False
    comment: str | None = Field(default=None, max_length=10000)
    clear_comment: bool = False
    expected_revision: int = Field(ge=1)


class CreateResearchSkillEvidenceRequest(BaseModel):
    start_frame: int
    end_frame_exclusive: int | None = None
    comment: str | None = Field(default=None, max_length=5000)
    expected_revision: int = Field(ge=1)


class UpdateResearchSkillEvidenceRequest(BaseModel):
    start_frame: int | None = None
    end_frame_exclusive: int | None = None
    clear_end_frame: bool = False
    comment: str | None = Field(default=None, max_length=5000)
    clear_comment: bool = False
    expected_revision: int = Field(ge=1)


class ResearchSkillMutationResponse(BaseModel):
    action: ResearchSkillMutationAction
    assessment: ResearchSkillAssessmentDetail
    changed_score_ids: list[int] = Field(default_factory=list)
    created_score_ids: list[int] = Field(default_factory=list)
    deleted_score_ids: list[int] = Field(default_factory=list)
    changed_evidence_ids: list[int] = Field(default_factory=list)
    created_evidence_ids: list[int] = Field(default_factory=list)
    deleted_evidence_ids: list[int] = Field(default_factory=list)


class ResearchSkillValidationIssue(BaseModel):
    issue_type: ResearchSkillValidationIssueType
    severity: ResearchSkillValidationSeverity
    message: str
    criterion_id: int | None = None
    score_id: int | None = None
    phase_segment_id: int | None = None
    evidence_id: int | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class ResearchSkillValidationIssueCounts(BaseModel):
    error: int
    warning: int
    info: int


class ResearchSkillValidationResponse(BaseModel):
    assessment_id: int
    revision: int
    status: ResearchSkillAssessmentStatus
    required_total: int
    required_completed: int
    completion_percent: float
    issue_counts: ResearchSkillValidationIssueCounts
    issues: list[ResearchSkillValidationIssue] = Field(default_factory=list)
    is_valid: bool
    can_submit: bool
    requires_warning_confirmation: bool


class SubmitResearchSkillAssessmentRequest(BaseModel):
    expected_revision: int = Field(ge=1)
    confirm_warnings: bool = False


class ReopenResearchSkillAssessmentRequest(BaseModel):
    expected_revision: int = Field(ge=1)


class ResearchSkillStatusMutationResponse(BaseModel):
    action: ResearchSkillStatusMutationAction
    assessment: ResearchSkillAssessmentDetail
    validation: ResearchSkillValidationResponse | None = None

File: frontend/src/types/researchSkill.ts
export type ResearchSkillRubricStatus = 'draft' | 'active' | 'archived'
export type ResearchSkillAssessmentStatus = 'draft' | 'submitted' | 'reviewed' | 'locked'
export type ResearchSkillCriterionScope = 'overall' | 'phase'
export type ResearchSkillScoreType = 'integer_scale' | 'number' | 'single_choice' | 'boolean' | 'text'
export type ResearchSkillTargetType = 'overall' | 'phase_segment'
export type ResearchSkillMutationAction =
  | 'created'
  | 'updated'
  | 'deleted'
  | 'unchanged'
  | 'evidence_created'
  | 'evidence_updated'
  | 'evidence_deleted'
export type ResearchSkillStatusMutationAction = 'submitted' | 'reopened'
export type ResearchSkillValidationSeverity = 'error' | 'warning' | 'info'
export type ResearchSkillValidationIssueType =
  | 'missing_required_score'
  | 'invalid_value'
  | 'na_not_allowed'
  | 'phase_score_without_segment'
  | 'segment_not_in_selected_phase_set'
  | 'criterion_not_applicable'
  | 'assessment_phase_set_missing'
  | 'assessment_phase_set_draft'
  | 'rubric_not_active'
  | 'inactive_criterion'
  | 'evidence_out_of_bounds'
  | 'incomplete_phase_criteria'

export type ResearchSkillPhaseLabel = {
  id: number
  key: string
  name: string
  color: string
}

export type ResearchSkillCriterion = {
  id: number
  rubric_id: number
  key: string
  name: string
  description: string | null
  scope: ResearchSkillCriterionScope
  score_type: ResearchSkillScoreType
  min_value: number | null
  max_value: number | null
  step: number | null
  options_json: Array<{ value: string; label: string }> | null
  required: boolean
  allow_na: boolean
  weight: number | null
  display_order: number
  is_active: boolean
  phase_label_ids: number[]
  phase_labels: ResearchSkillPhaseLabel[]
  created_at: string
  updated_at: string
}

export type ResearchSkillCriterionPhaseLabel = {
  criterion_id: number
  phase_label_id: number
}

export type ResearchSkillRubricSummary = {
  id: number
  name: string
  version: number
  description: string | null
  status: ResearchSkillRubricStatus
  phase_protocol_id: number | null
  created_by_id: number | null
  criterion_count: number
  created_at: string
  updated_at: string
}

export type ResearchSkillRubricDetail = ResearchSkillRubricSummary & {
  criteria: ResearchSkillCriterion[]
}

export type ResearchSkillVideoSummary = {
  id: number
  name: string
  fps: number | null
  frame_count: number
  duration_ms: number | null
}

export type ResearchSkillPhaseSegment = {
  id: number
  phase_label_id: number
  phase_key: string
  phase_name: string
  start_frame: number
  end_frame_exclusive: number | null
}

export type ResearchSkillPhaseAnnotationSet = {
  id: number
  protocol_id: number
  status: string
  revision: number
  segments: ResearchSkillPhaseSegment[]
}

export type ResearchSkillEvidence = {
  id: number
  skill_score_id: number
  start_frame: number
  end_frame_exclusive: number | null
  comment: string | null
  created_at: string
  updated_at: string
}

export type ResearchSkillScore = {
  id: number
  assessment_id: number
  criterion_id: number
  criterion_key: string
  criterion_name: string
  scope: ResearchSkillCriterionScope
  score_type: ResearchSkillScoreType
  target_key: string
  phase_segment_id: number | null
  value: unknown
  is_na: boolean
  comment: string | null
  evidence: ResearchSkillEvidence[]
  created_at: string
  updated_at: string
}

export type ResearchSkillCompletion = {
  required_total: number
  required_completed: number
  overall_required_total: number
  overall_required_completed: number
  phase_required_total: number
  phase_required_completed: number
  completion_percent: number
}

export type ResearchSkillAssessmentSummary = {
  id: number
  video_id: number
  rubric_id: number
  rater_id: number
  phase_annotation_set_id: number | null
  status: ResearchSkillAssessmentStatus
  revision: number
  overall_comment: string | null
  submitted_at: string | null
  reviewed_at: string | null
  locked_at: string | null
  created_at: string
  updated_at: string
  rubric_name: string
  rubric_version: number
  rater_username: string
  score_count: number
}

export type ResearchSkillAssessmentDetail = ResearchSkillAssessmentSummary & {
  video: ResearchSkillVideoSummary
  rubric: ResearchSkillRubricDetail
  phase_annotation_set: ResearchSkillPhaseAnnotationSet | null
  scores: ResearchSkillScore[]
  completion: ResearchSkillCompletion
}

export type ResearchSkillValidationIssue = {
  issue_type: ResearchSkillValidationIssueType
  severity: ResearchSkillValidationSeverity
  message: string
  criterion_id: number | null
  score_id: number | null
  phase_segment_id: number | null
  evidence_id: number | null
  details: Record<string, unknown>
}

export type ResearchSkillValidationResponse = {
  assessment_id: number
  revision: number
  status: ResearchSkillAssessmentStatus
  required_total: number
  required_completed: number
  completion_percent: number
  issue_counts: { error: number; warning: number; info: number }
  issues: ResearchSkillValidationIssue[]
  is_valid: boolean
  can_submit: boolean
  requires_warning_confirmation: boolean
}

export type CreateSkillRubricRequest = {
  name: string
  version?: number
  description?: string | null
  phase_protocol_id?: number | null
  username?: string | null
}

export type UpdateSkillRubricRequest = {
  name?: string | null
  description?: string | null
  phase_protocol_id?: number | null
  clear_phase_protocol?: boolean
}

export type CloneSkillRubricRequest = {
  name?: string | null
  description?: string | null
}

export type CreateSkillCriterionRequest = {
  key: string
  name: string
  description?: string | null
  scope: ResearchSkillCriterionScope
  score_type: ResearchSkillScoreType
  min_value?: number | null
  max_value?: number | null
  step?: number | null
  options_json?: Array<{ value: string; label: string }> | null
  required?: boolean
  allow_na?: boolean
  weight?: number | null
  display_order: number
  is_active?: boolean
  phase_label_ids?: number[]
}

export type UpdateSkillCriterionRequest = Partial<CreateSkillCriterionRequest> & {
  clear_weight?: boolean
}

export type CreateSkillAssessmentRequest = {
  rubric_id: number
  username: string
  phase_annotation_set_id?: number | null
}

export type UpdateSkillAssessmentRequest = {
  overall_comment?: string | null
  clear_overall_comment?: boolean
  phase_annotation_set_id?: number | null
  clear_phase_annotation_set?: boolean
  expected_revision: number
}

export type UpsertSkillScoreRequest = {
  target_type: ResearchSkillTargetType
  phase_segment_id?: number | null
  value?: unknown
  is_na?: boolean
  comment?: string | null
  clear_comment?: boolean
  expected_revision: number
}

export type CreateSkillEvidenceRequest = {
  start_frame: number
  end_frame_exclusive?: number | null
  comment?: string | null
  expected_revision: number
}

export type UpdateSkillEvidenceRequest = {
  start_frame?: number | null
  end_frame_exclusive?: number | null
  clear_end_frame?: boolean
  comment?: string | null
  clear_comment?: boolean
  expected_revision: number
}

export type SubmitSkillAssessmentRequest = {
  expected_revision: number
  confirm_warnings?: boolean
}

export type ReopenSkillAssessmentRequest = {
  expected_revision: number
}

export type ResearchSkillMutationResponse = {
  action: ResearchSkillMutationAction
  assessment: ResearchSkillAssessmentDetail
  changed_score_ids: number[]
  created_score_ids: number[]
  deleted_score_ids: number[]
  changed_evidence_ids: number[]
  created_evidence_ids: number[]
  deleted_evidence_ids: number[]
}

export type ResearchSkillStatusMutationResponse = {
  action: ResearchSkillStatusMutationAction
  assessment: ResearchSkillAssessmentDetail
  validation: ResearchSkillValidationResponse | null
}

export type CreateSkillAssessmentResponse = {
  created: boolean
  assessment: ResearchSkillAssessmentDetail
}

export type ResearchSkillConflictDetail = {
  message: string
  current_revision?: number | null
}

export type ResearchSkillValidationErrorDetail = {
  message: string
  validation?: ResearchSkillValidationResponse
}
